src/NServiceBus.Core/ObjectBuilder/Common/CommonObjectBuilder.cs

```diff
@@ -3,19 +3,50 @@ namespace NServiceBus
     using System;
     using System.Collections.Generic;
     using System.Linq;
+#if NET452
+    using System.Runtime.Remoting.Messaging;
+#endif
+#if NETCOREAPP2_0
+    using System.Threading;
+#endif
     using ObjectBuilder;
     using ObjectBuilder.Common;
 
     class CommonObjectBuilder : IBuilder, IConfigureComponents
     {
+#if NET452
+        readonly string rootBuilderKey = Guid.NewGuid().ToString();
+        public IBuilder CurrentBuilder
+        {
+            get
+            {
+                var logicalGetData = (IBuilder)CallContext.LogicalGetData(rootBuilderKey);
+                return logicalGetData ?? this;
+            }
+            set
+            {
+                CallContext.LogicalSetData(rootBuilderKey, value);
+            }
+        }
+#endif
+#if NETCOREAPP2_0
+        AsyncLocal<IBuilder> childBuilder = new AsyncLocal<IBuilder>();
+        public IBuilder CurrentBuilder
+        {
+            get { return childBuilder.Value ?? this; }
+            set { childBuilder.Value = value; }
+        }
+#endif
         public CommonObjectBuilder(IContainer container)
         {
             this.container = container;
         }
 
         public IBuilder CreateChildBuilder()
         {
-            return new CommonObjectBuilder(container.BuildChildContainer());
+            var childBuilder = new CommonObjectBuilder(container.BuildChildContainer());
```

Comment:
yeah might need additional logic to clear the assigned child container once that is disposed. Will write a test to verify.